File: 3445-lexicographically-minimum-string-after-removing-stars/3445-lexicographically-minimum-string-after-removing-stars.java
class Solution {
    public String clearStars(String s) {
        int n = s.length();
        PriorityQueue<Character> pq = new PriorityQueue<>();
        HashMap<Character, Deque<Integer>> mp = new HashMap<>();
        boolean[] check = new boolean[n];
        Arrays.fill(check, true);

        for(int i = 0; i<n; i++) {
            char c = s.charAt(i);
            if(c == '*') {
                char temp = pq.poll();
                int idx = mp.get(temp).removeLast();
                check[i] = false;
                check[idx] = false;
            } else {
                pq.offer(c);
                mp.putIfAbsent(c, new ArrayDeque<>());
                mp.get(c).add(i);
            }
        }

        StringBuilder ans = new StringBuilder();
        for(int i = 0; i<n; i++) {
            if(check[i]) ans.append(s.charAt(i));
        }

        return ans.toString();
    }
}

File: 3445-lexicographically-minimum-string-after-removing-stars/3445-lexicographically-minimum-string-after-removing-stars.cpp
class Solution {
public:
    string clearStars(string s) {
        int n = s.size();
        priority_queue<char, vector<char>, greater<char>> pq;
        unordered_map<char, vector<int>> mp;
        vector<bool> check(n, true);

        for(int i = 0; i<n; i++) {
            if(s[i] == '*') {
                char temp = pq.top();
                pq.pop();
                int idx = mp[temp].back();
                mp[temp].pop_back();
                check[i] = false;
                check[idx] = false;
            } else {
                pq.push(s[i]);
                mp[s[i]].push_back(i);
            }
        }

        string ans = "";
        for(int i = 0; i<n; i++) {
            if(check[i]) ans += s[i];
        }

        return ans;
    }
};